types/src/bytesrepr.rs

```diff
@@ -772,6 +836,213 @@ impl<T: FromBytes, E: FromBytes> FromBytes for Result<T, E> {
+            FromBytes::from_bytes(bytes)?;
+        Ok((octets.into(), remainder))
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for IpAddr {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        match self {
+            IpAddr::V4(_) => U8_SERIALIZED_LENGTH + IPV4_SERIALIZED_LENGTH,
+            IpAddr::V6(_) => U8_SERIALIZED_LENGTH + IPV6_SERIALIZED_LENGTH,
+        }
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        match self {
+            IpAddr::V4(addr) => {
+                writer.push(IP_V4_TAG);
+                addr.write_bytes(writer)
+            }
+            IpAddr::V6(addr) => {
+                writer.push(IP_V6_TAG);
+                addr.write_bytes(writer)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl FromBytes for IpAddr {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (tag, remainder) = u8::from_bytes(bytes)?;
+        match tag {
+            IP_V4_TAG => {
+                let (addr, rem) = Ipv4Addr::from_bytes(remainder)?;
+                Ok((Self::V4(addr), rem))
+            }
+            IP_V6_TAG => {
+                let (addr, rem) = Ipv6Addr::from_bytes(remainder)?;
+                Ok((Self::V6(addr), rem))
+            }
+            _ => Err(Error::Formatting),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for SocketAddrV4 {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U16_SERIALIZED_LENGTH + IPV4_SERIALIZED_LENGTH
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.port().write_bytes(writer)?;
+        self.ip().write_bytes(writer)
+    }
+}
+
+#[cfg(feature = "std")]
+impl FromBytes for SocketAddrV4 {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (port, remainder): (u16, &[u8]) = FromBytes::from_bytes(bytes)?;
+        let (addr, rem): (Ipv4Addr, &[u8]) = FromBytes::from_bytes(remainder)?;
+        Ok((Self::new(addr, port), rem))
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for SocketAddrV6 {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U16_SERIALIZED_LENGTH
+            + U32_SERIALIZED_LENGTH
+            + U32_SERIALIZED_LENGTH
+            + IPV6_SERIALIZED_LENGTH
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.port().write_bytes(writer)?;
+        self.flowinfo().write_bytes(writer)?;
+        self.scope_id().write_bytes(writer)?;
```

Comment:
I am unsure whether we want to persist the `scope_id` and `flowinfo` parts. On the one hand, they are part of the serialization and naively I would expect for them to roundtrip correctly, on the other hand their presence on an IPv6 address seems more like an implementation detail of the underlying C struct.

`scope_id` is only valid on the local machine, as it names a local interface. Should we ever support IPv6, we probably would not be sending this across the network as a gossiped address.

`flowinfo` has meaning outside the local machine, but I wonder whether it should be persisted as well. If used, it is typically filled before sending.

My limited intuition would also opt for an "ip, port" serialization. Maybe someone with more comprehensive IPv6 experience can help out here?

Reply:
My personal preference here would be to keep `bytesrepr` implementations as strictly (de)serializing structures completely (no losses) and not actually reason about what fields are relevant or not in the implementation. I think that kind of reasoning should be done at a higher level, kind of how `serde` does it with `bincode`. This is to keep the `bytesrepr` functionality consistent and to avoid any weird corner cases in its behavior. In the absence of a higher level implementation, you would need to either:
- Create new structs which serve as a schema containing exactly what you need (de)serialized. In this particular case, we would have a
```
CustomSocketAddrV6 {
    ip: IpV6Addr,
    port: u16,
}
```
- Reason about the fields pre-serialization/post-deserialization.

Reply:
That's something I can certainly support. I kinda blame the actual types tacking on these extra fields :/

My reasoning is that should we ever use a serialized representation of this in a hashed way, we'd have to zero some of the fields.